src/hotspot/share/opto/memnode.cpp

```diff
@@ -1561,22 +1589,67 @@ bool LoadNode::can_split_through_phi_base(PhaseGVN* phase) {
     base = base->in(1);
   }
 
-  if (req() > 3 || base == nullptr || !base->is_Phi()) {
+   if (req() > 3 || !get_region_to_split_through_phi(base, mem)) {
     return false;
   }
 
-  if (!mem->is_Phi()) {
-    if (!MemNode::all_controls_dominate(mem, base->in(0)))
-      return false;
-  } else if (base->in(0) != mem->in(0)) {
-    if (!MemNode::all_controls_dominate(mem, base->in(0))) {
-      return false;
+  if (nested) {
+    Node *parent_phi = nullptr;
+    // find parent phi node for nestedphi
+    for (uint i = 1; i < base->req(); i++) {
+      if (base->in(i)->is_Phi()) {
+        parent_phi = base->in(i);
+        Node *mem_node_for_load_after_opt = get_memory_node_for_nested_phi_after_optimization(phase, base, parent_phi);
+        if (!mem_node_for_load_after_opt || !get_region_to_split_through_phi(parent_phi, mem_node_for_load_after_opt)) {
+          return false;
+        }
+      }
     }
   }
 
   return true;
 }
 
+//------------------------------get_memory_node_for_nested_phi_after_optimization------------------------------
+// Given a base phi node and its parent phi node, this function pretends that optimizations have occurred on load fields and returns the optimized memory node.
+// Note that this function doesn't perform any optimizations itself.
+// If optimization is impossible, it returns nullptr.
+Node* LoadNode::get_memory_node_for_nested_phi_after_optimization(PhaseGVN* phase, Node *basephi, Node* base_parentphi) {
+  Node* mem        = in(Memory);
+  Node *address    = in(Address);
+  assert(basephi != nullptr && basephi->is_Phi(), "sanity");
+  assert(base_parentphi != nullptr && base_parentphi->is_Phi(), "sanity");
+  Node* region = get_region_to_split_through_phi(basephi, mem);
+  if (region == nullptr && basephi->in(0) != mem->in(0)) {
```

Comment:
This looks similar to some of the code in get_region_to_split_through_phi. Is this redundant logic? If not, can the logic here be moved into get_region_to_split_through_phi?